### DISCUSSION TITLE
[Bug] dispose error when using provideSignatureHelp when using model

### DISCUSSION THREAD

**Discussion by fabiojpoli** [microsoft/monaco-editor#3688] [Q&A]
### Reproducible in vscode.dev or in VS Code Desktop?

- [X] Not reproducible in [vscode.dev](https://vscode.dev) or VS Code Desktop

### Reproducible in the monaco editor playground?

- [ ] Not reproducible in [the monaco editor playground](https://microsoft.github.io/monaco-editor/playground.html)

### Monaco Editor Playground Code

```typescript
var code = "function hello() {\n\talert('Hello world!');\n\tmyFn\n}";
var model = monaco.editor.createModel(code, 'javascript');

monaco.editor.create(document.getElementById('container'), {
	model: model
});

monaco.languages.registerSignatureHelpProvider('javascript', {
    signatureHelpTriggerCharacters: ['(', ','],
    provideSignatureHelp: function(model, position) {
      return {
         value: {
            activeSignature: 0,
            activeParameter: 0,
            signatures: [{
               label: 'method',
               documentation: 'this method does blah',
               parameters: [{
                  label: 'ParamInfo1',
                  documentation: 'this param does blah1'
               }, {
                  label: 'ParamInfo2',
                  documentation: 'this param does blah2'
               }]
            }]
         }
      };
   }
});
```

### Reproduction Steps

Type `(p1,p2` after myFn

### Actual (Problematic) Behavior

Console error:
errors.ts:26 Uncaught Error: h.dispose is not a function

TypeError: h.dispose is not a function
    at set value [as value] (lifecycle.ts:289:16)
    at g.<anonymous> (parameterHintsModel.ts:221:35)
    at Generator.next (<anonymous>)
    at b (css.ts:80:37)
    at errors.ts:26:12

### Expected Behavior

No Console errors

### Additional Context

Additional Question:
How to implement `provideSignatureHelp` correctly? There is no example using it and documentation doesn't match with return format.

I just need to see this when type "(" for first parameter:
![Screen Shot 2023-01-12 at 2 38 57 PM](https://user-images.githubusercontent.com/663908/212176182-dcf15ac9-475a-4836-a73f-a9f456d2ec9d.png)

And this for second parameter when type ",":
![Screen Shot 2023-01-12 at 2 39 08 PM](https://user-images.githubusercontent.com/663908/212176230-09e561e0-25f0-4d72-8b1f-a08b2239aefe.png)

With highlight on each parameter. Besides the bug, could you please help me to implement this?

Thank you!

**Comment by tamayika** [comment]
I don't know the historical reason but you need to implement dispose. (while VSCode extension API does not)

```js
var code = "function hello() {\n\talert('Hello world!');\n\tmyFn\n}";
var model = monaco.editor.createModel(code, 'javascript');

monaco.editor.create(document.getElementById('container'), {
	model: model
});

monaco.languages.registerSignatureHelpProvider('javascript', {
    signatureHelpTriggerCharacters: ['(', ','],
    provideSignatureHelp: function(model, position) {
      return {
         value: {
            activeSignature: 0,
            activeParameter: 0, // you need to change which parameter is typing
            signatures: [{
               label: 'blah(blah1, blah2)', // siagnature label must be entire function signature
               documentation: 'this method does blah',
               parameters: [{
                  label: 'blah1', // parameter label must be partial of signature label
                  documentation: 'this param does blah1',
               }, {
                  label: 'blah2', // parameter label must be partial of signature label
                  documentation: 'this param does blah2'
               }]
            }]
         },
         dispose: () => {},
      };
   }
});
```

**Comment by fabiojpoli** [comment]
Thank you @tamayika!

The hint is not closing automatically after type ")". How could I do that manually to fix this problem? Thank you!

**Comment by fabiojpoli** [comment]
For now my workaround is:
```js
provideSignatureHelp: function(model, position) {
   var me = this,
      currentChar = model.getLineContent(position.lineNumber)[position.column - 2];

   // if is typing ")" just close the parameters hint
   if (currentChar === ')') {
      return {
         dispose: () => {},
         value: {}
      };
   }
   ...
}
```

**Comment by tamayika** [comment]
Yes, you need to return empty result when out of call expression.
Generally, you need AST parser because you can not determine current text is comment or string literal or call expression.
With AST, you can know argument index and out of call expression.
